### ISSUE TITLE
Problems with while(x--)

### ISSUE THREAD

**Issue by a-turko** [MPI-SWS/genmc#42]
If the program being verified uses the value of x-- directly as a condition, genmc raises an error. However, if the decrement operation and the condition are separated, the code is correctly verified by genmc.
This is somewhat suspicious, because both approaches are equivalent (the decremented variable is local).

Consider the following example:

```C
atomic_int data;
void *thread()
{
#ifdef USE_DEC_WHILE1
    int lmt = N_ITER;
    while(lmt--) data++;
}
#elif USE_LONG_WHILE
    int lmt = N_ITER;
    while(lmt) {
        data++;
        lmt--;
    }
#endif
```

The first variant produces an error and the other does not:
```
user@8976719904df:~/cds/test/scq$ genmc --version
GenMC (https://plv.mpi-sws.org/genmc):
  GenMC v0.8
  Built with LLVM 8.0.0 (RelWithDebInfo)
user@8976719904df:~/cds/test/scq$ genmc -- -DUSE_DEC_WHILE1 minimal.c 
Only PHI nodes may reference their own value!
  %dec = add nsw i32 %dec, -1, !dbg !50
genmc: LLVMModule.cpp:202: bool LLVMModule::transformLLVMModule(llvm::Module &, ModuleInfo &, const std::shared_ptr<const Config> &): Assertion `!llvm::verifyModule(mod, &llvm::dbgs())' failed.
Aborted (core dumped)
user@8976719904df:~/cds/test/scq$ genmc -- -DUSE_LONG_WHILE minimal.c 
No errors were detected.
Number of complete executions explored: 1
Total wall-clock time: 0.01s
```

This is the full code of the program producing the issue (with additional examples):
```C

#include <stdlib.h>
#include <pthread.h>
#include <stdatomic.h>
#include <assert.h>

atomic_int data;

#define N_ITER 2

void *thread(void *unused)
{	
#ifdef USE_DEC_WHILE1
	int lmt = N_ITER;
	while (lmt--) data++;

	// error

#elif USE_DEC_WHILE2
	int lmt = N_ITER + 1;
	while (--lmt) data++;

	// error

#elif USE_LONG_WHILE
	int lmt = N_ITER;
	while (lmt) {
		data++;
		lmt--;
	}

	// ok

#elif USE_INC_WHILE
	int lmt = 0;
	while (lmt++ < N_ITER) data++;

	// error

#elif USE_FOR
	for (int i = 0; i < N_ITER; i++) data++;

	// ok

#elif USE_IF
	int lmt = N_ITER;
	while (1) {
		if (lmt--) data++;
		else break;
	}

	// error

#else
	DO NOT COMPILE -- missing variant
#endif

	return NULL;
}

int main()
{
	pthread_t t;
	data = 0;

	pthread_create(&t, NULL, thread, NULL);

	pthread_join(t, NULL);

	assert(data == N_ITER);
}

```

**Comment by michaliskok** [comment]
Thanks for the report and sorry for the late reply. 

This is a bug in one of the transformations employed by GenMC; I will push a fix in the next release. In the meantime, you can use `-disable-loop-jump-threading` to disable the transformation.

Michalis

**Comment by michaliskok** [comment]
Fixed in v0.9.
